### DISCUSSION TITLE
Vavr extension

### DISCUSSION THREAD

**Discussion by rfuerst87** [instancio/instancio#825] [Q&A]
Hi @armandino 

I created an issue two days ago asking for supporting Vavr collections. As I had no better plans today I thought to give it a try myself.  You can find my progress in https://github.com/movementsciences/instancio-vavr/.

I have a few questions and problems that I hope you can help me with:

1. As Vavr collections do not implement `java.util.Collection` I assume I have to register them as container types in `InternalContainerFactoryProvider`. Is that correct?
2. What is `InternalContainerFactoryProvider#getMappingFunction` used for?
3. As I just mentioned Vavr collections do not implement `java.util.Collection`. So using `CollectionGenerator` as done in `instancio-guava` does not work. I therefore came up with [`VavrSeqGenerator`](https://github.com/movementsciences/instancio-vavr/blob/feature/initial-implementation/src/main/java/ch/movementsciences/instancio/vavr/internal/generator/VavrSeqGenerator.java). Problem is that it fails for all types that is not the default `List` implementation. All test in [`VavrSupportedSeqTest`](https://github.com/movementsciences/instancio-vavr/blob/feature/initial-implementation/src/test/java/ch/movementsciences/instancio/vavr/VavrSupportedSeqTest.java) fail. Turns out that java collections use a dedicated [`CollectionNodeHandler`](https://github.com/instancio/instancio/blob/main/instancio-core/src/main/java/org/instancio/internal/handlers/CollectionNodeHandler.java) that sets the subtype. [`UsingGeneratorResolverHandler`](https://github.com/instancio/instancio/blob/main/instancio-core/src/main/java/org/instancio/internal/handlers/UsingGeneratorResolverHandler.java) that is used for SPI provided generators does not do that. Would it be possible to set the subtype in `UsingGeneratorResolverHandler` if the generator implements the `SubtypeGeneratorSpec`? Or am I doing this entirely wrong?

**Comment by armandino** [comment]
@rfuerst87 I had a quick at the project. Great start! It's been a while since I added the guava module, but hopefully my memory serves me well and I answer your questions correctly.

1. Yes. Basically marking it as a container means that the object needs be populated using a specific method like `add(T)`, `put(K, V)`, `Optional.of()`, etc. In addition, the container's method may need to be called more than once (e.g. for collection-like classes). The generator for the container needs to provide an `InternalContainerHint` to specify the "add" method and container size, e.g.

```java
InternalContainerHint.builder()
    // how many times the "add" function should be called
    .generateEntries(numberOfEntriesToAdd)
    // method for adding to the container, e.g. Seq
    .addFunction((Seq<T> seq, Object... args) -> seq.append((T) args[0]))
    // Or for Map
    //.addFunction((Map<K, V> seq, Object... args) -> map.put((K) args[0], (V) args[1]))
    .build();
```

The size of the `args` array depends on how many type variables the container's class declares. E.g. `Map<K, V>` has two types `K` and `V`.  Instancio will generate an object of type `K` and `V` and provide the values via the `args` array.

2. The `InternalContainerFactoryProvider#getMappingFunction` is used for converting a collection from one type to another. For example, Guava's `ImmutableList` can't be generated using a regular generator because its `add()` method throws an error. Using `getMappingFunction` we can generate a regular `ArrayList` and covert to `ImmutableList` using `ImmutableList.copyOf()`.

3. I'm not sure if we can leverage `CollectionGenerator` because, as you mentioned, VAVR collections aren't based on `java.util.Collection`. For this reason, it may not be possible to use the built-in `gen.collection()`method. So the subtype functionality would probably need to be implemented by the `VavrSeqGenerator`.

As I mentioned, the internal API for generating containers isn't super intuitive :) I can take a closer look into it once I have some free time. Let me know if you have any other questions in the meantime.

**Reply by rfuerst87** [reply]
Many thanks for the quick response @armandino!

> The InternalContainerFactoryProvider#getMappingFunction is used for converting a collection from one type to another. For example, Guava's ImmutableList can't be generated using a regular generator because its add() method throws an error. Using getMappingFunction we can generate a regular ArrayList and covert to ImmutableList using ImmutableList.copyOf().

I see. So I could probably use that to replace the subtype. My other idea was to explicitly set the subtype in the generator in `UsingGeneratorResolverHandler` if the generator implements `SubtypeGeneratorSpec`. That way it works exactly the same as in `CollectionNodeHandler`. To me this looks like the cleaner solution and gives all user provided generators the ability to decide on the subtype. Replacing generated objects isn't needed in that case. WDYT?

> I'm not sure if we can leverage CollectionGenerator because, as you mentioned, VAVR collections aren't based on java.util.Collection. For this reason, it may not be possible to use the built-in gen.collection()method. So the subtype functionality would probably need to be implemented by the VavrSeqGenerator.

`CollectionGenerator` most doesn't work for vavr collections. At least I couldn't figure out a way. `VavrSeqGenerator` implements the subtype functionality, but it isn't used by the `UsingGeneratorResolverHandler` as mentioned before. This is why `VavrListTest` work but `VavrSupportedSeqTest` don't.

**Reply by armandino** [reply]
> I see. So I could probably use that to replace the subtype. My other idea was to explicitly set the subtype in the generator in `UsingGeneratorResolverHandler` if the generator implements `SubtypeGeneratorSpec`. That way it works exactly the same as in `CollectionNodeHandler`. To me this looks like the cleaner solution and gives all user provided generators the ability to decide on the subtype. Replacing generated objects isn't needed in that case. WDYT?

I'm not sure what the side-effects of that will be. Have you tried it to see if existing tests still pass?

**Reply by armandino** [reply]
@rfuerst87 I experimented a bit and it looks like this PR should make the current tests pass (I removed tests that use `gen.collection()` since it can't be used with VAVR).

https://github.com/movementsciences/instancio-vavr/pull/1

**Reply by rfuerst87** [reply]
Hi @armandino, thx! This was actually a very silly mistake by me. The collection generator wasn't meant to be used in the test. I cleaned up stuff a little and added a test to make things easier for discussion. 

- [`generatorSpecSubtype()`](https://github.com/movementsciences/instancio-vavr/blob/6b3fc9b2972b6403c495edf527dba3b042deb1e4/src/test/java/ch/movementsciences/instancio/vavr/ListTest.java#L63) tests whether the default implementation in the `SeqGenerator` is used.
- [`generatorSpecSubtypeNonDefault()`](https://github.com/movementsciences/instancio-vavr/blob/6b3fc9b2972b6403c495edf527dba3b042deb1e4/src/test/java/ch/movementsciences/instancio/vavr/ListTest.java#L72]) tests whether any other subtype is used. 

The second test always fails as this information is never propagated to the generator. As far as I understand this subtype is also not available in `InternalContainerFactoryProvider#getMappingFunction`. I therefore tried to update `UsingGeneratorResolverHandler` to set the subtype directly if the generator implements `SubtypeGeneratorSpec`. However this seems to break existing tests. The only remaining solution seems to be create a generator per vavr collection type. Am I missing something?

**Comment by armandino** [comment]
@rfuerst87 I think your initial approach was on the right track (generating a `java.util.Collection` and converting it via `getMappingFunction()` to an implementation of `Seq`).

Please see the updated PR https://github.com/movementsciences/instancio-vavr/pull/2 with passing tests.

**Reply by rfuerst87** [reply]
Awesome, thank you very much! I will take a closer look as soon as I have time.

**Reply by rfuerst87** [reply]
I finally found some time to take a look on your PR. The piece I was missing is to use `InternalGeneratorHint` if a subtype is set in my generator. Basically with that you solved my pain points. From now it should be straight forward. I will ping you again when I have some more time to work on it.

**Reply by armandino** [reply]
@rfuerst87 sure thing, let me know if you run into any other issues.

**Comment by aalbiach** [comment]
Hi, I have created a VAVR module in a fork of instancio, I didn't know that @rfuerst87 had started to develop the VAVR extension.
I guess approaching this development as an extension instead of a module within the same project is more appropriate, so I won't create the PR for now, unless you think otherwise.
@rfuerst87 in case it helps you, you can take a look at the code I created here https://github.com/aalbiach/instancio/tree/feat/create-vavr-module/instancio-vavr

**Reply by armandino** [reply]
@aalbiach - thank you for putting it together, much appreciated. We discussed adding a module vs extension in #823. Creating an extension seemed liked a better option, at least for the time being (the linked issue has more details on this). @rfuerst87 has kindly agreed to set up a project. Not sure how far along it is, but it's probably best joining efforts to reuse any missing functionality and tests.

**Reply by rfuerst87** [reply]
Hi @aalbiach, cool to see there is more interest in a Vavr extension. In https://github.com/movementsciences/instancio-vavr/tree/feature/initial-implementation the hard parts are solved I think. I had some troubles with subtype support, but with the help of @armandino I've got it to work. What's left is the actual implementation. That is:

- Support vor `CharSeq`. I'm not yet sure how to deal with it
- Support for `Set` and `Map`. Needs to be implemented... Should be straight forward
- Support for for values (`Option`, `Try`, `Either`, `Future` and `Validation`)

I don't have much time to work on the project... Just a few hours here and then. If you have time, we could indeed join forces. Let me know whether you're interested.

**Reply by aalbiach** [reply]
Hi @rfuerst87, I am willing to collaborate with you in the development of this extension. I have some more time to dedicate to this project and I am motivated to create a Vavr extension that is useful and easy to use.
I will review the current code and post some PR's for you to review.
